### PR TITLE
feat(sdk): native fetch

### PR DESCRIPTION
**Pull request by tsuf239**
fixes https://github.com/winglang/wing/issues/2169

* adding a static `Http` class to `wingsdk.util`
* replacing the use of external fetch in `api_path_vars.w` file
* removing   `api_path_vars.js` -contains the js fetch

left todo:
- [x] adding more tests
- [ ] Updating docs (prefer to do it after the implementation is reviewed)

## Checklist

- [x] Title matches [Winglang's style guide](https://docs.winglang.io/contributors/pull_requests#how-are-pull-request-titles-formatted)
- [x] Description explains motivation and solution
- [ ] Tests added (always)
- [ ] Docs updated (only required for features)
- [ ] Added `pr/e2e-full` label if this feature requires end-to-end testing

*By submitting this pull request, I confirm that my contribution is made under the terms of the [Monada Contribution License](https://docs.winglang.io/terms-and-policies/contribution-license.html)*.

### CODE REVIEW

File: examples/tests/valid/api_path_vars.w [hunk 1]
```diff
@@ -1,10 +1,8 @@
 bring cloud;
+bring util;
```
Comment (eladb) [inline examples/tests/valid/api_path_vars.w:2]: This should be `bring http`, not `bring util`.

File: examples/tests/valid/api_path_vars.w [hunk 2]
```diff
@@ -15,13 +13,13 @@ let handler = inflight (req: cloud.ApiRequest): cloud.ApiResponse => {
 
 api.get("/users/{name}", handler);
 
-let f = new Fetch();
 
 test "test" {
   let username = "tsuf";
-  // TODO: change f.get to static when possible
-  let res = f.get("${api.url}/users/${username}");
+  let res: util.Response = util.Http.get("${api.url}/users/${username}");
```
Comment (eladb) [inline examples/tests/valid/api_path_vars.w:19]:
```suggestion
  let res: util.Response = http.get("${api.url}/users/${username}");
```

File: libs/wingsdk/src/util/fetch.ts [hunk 3]
```diff
@@ -0,0 +1,260 @@
+import { Code, InflightClient } from "../core";
```
Comment (eladb) [inline libs/wingsdk/src/util/fetch.ts:1]: This file should be under `src/http/fetch.ts` inside a class named `Util`:

`src/http/fetch.ts`:

```js
export class Util {
  public async get(...);
  public async fetch(...);
}
```

The `Util` class name is [special] and will automatically turn into a "global namespace" for the submodule.

[special]: https://github.com/winglang/wing/blob/a69af8e0451101770dda4d240d9521c4cc31191a/libs/wingc/src/type_check.rs#L3136

File: libs/wingsdk/src/util/fetch.ts [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+  headers: {},
+  cache: RequestCache.DEFAULT,
+  redirect: RequestRedirect.FOLLOW,
+};
+/**
+ * the Http class is used for calling different HTTP methods and requesting and sending information online,
+ *  as well as testing public accessible resources
+ * @inflight
+ */
+export class Http {
```
Comment (eladb) [inline libs/wingsdk/src/util/fetch.ts:150]:
```suggestion
export class Util {
```

**Review by eladb — CHANGES_REQUESTED** [review]

**Comment by tsuf239** [comment]
It seems like the body needs to be parsed twice (JSON.parse(JSON.parse(response.body))) in order to get its value into Json, in tf-aws, will fix in the current or following PR 🙂

**Comment by tsuf239** [comment]
> It seems like the body needs to be parsed twice (JSON.parse(JSON.parse(response.body))) in order to get its value into Json, in tf-aws, will fix in the current or following PR 🙂

Fixed. found the other stringifying method :)

**Comment by MarkMcCulloh** [comment]
Something to note: we should raise the required node version in our packages/docs from just "18" to "8.13.0", since fetch was experimental until that version. Otherwise users will get a scary warning about it.

**Comment by eladb** [comment]
> Something to note: we should raise the required node version in our packages/docs from just "18" to "8.13.0", since fetch was experimental until that version. Otherwise users will get a scary warning about it.

That's a great point. Maybe even raise to the latest LTS 18.16.1?

File: libs/wingsdk/test/http/http.test.ts [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+  Util as Http,
+  HttpMethods,
+  Response,
+  RequestCache,
+  RequestRedirect,
+} from "../../src/http/http";
+
+let defaultOptions = {
+  method: HttpMethods.GET,
+  headers: {},
```
Comment (eladb) [inline libs/wingsdk/test/http/http.test.ts:12]: Can we please write this test with wing instead of typescript?

File: libs/wingsdk/src/http/http.ts [hunk 6]
```diff
@@ -0,0 +1,259 @@
...
+  /**
+   * The cache mode you want to use for the request.
+   */
+  readonly cache?: RequestCache;
+  /**
+   * he redirect mode to use: follow, error. The default is follow.
+   */
+  readonly redirect?: RequestRedirect;
+  /**
+   * A string specifying "no-referrer", client, or a URL. The default is "about:client".
```
Comment (eladb) [inline libs/wingsdk/src/http/http.ts:107]: Is that really the default?

**Review by eladb — CHANGES_REQUESTED** [review]

File: libs/wingsdk/src/http/http.ts [hunk 7]
```diff
@@ -0,0 +1,259 @@
+import { Code, InflightClient } from "../core";
+
+/**
+ * RequestCache-
+ * The cache read-only property that contains the cache mode of the request.
+ * It controls how the request will interact with the browser's HTTP cache.
+
```
Comment (Chriscbr) [inline libs/wingsdk/src/http/http.ts:7]:
```suggestion
 * The cache mode of the request.
 * It controls how a request will interact with the system's HTTP cache.
```

File: libs/wingsdk/src/http/http.ts [hunk 8]
```diff
@@ -0,0 +1,259 @@
...
+ */
+export enum RequestCache {
+  /**
+   * The browser looks for a matching request in its HTTP cache.
+   * If there is a match and it is fresh, it will be returned from the cache.
```
[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingsdk/src/http/http.ts:12]: Sure thing. I just did some testing, and for showing bullets in our docs, the list needs to (1) start with an empty line before it, and (2) each start with a `*` character. For example:

```js
/**
 * Here is my description, blah blah
 *
 * * This is my first bullet point. The sentence may go on
 * for multiple lines.
 * * This is my second bullet point. Our documentation can
 * go on forever!
 *
 * More content can follow if I want.
 */
```

If we follow these rules then our `jsii-docgen` tool seems to not get confused. :)

@tsuf239 If you try running `npm run compile` and `npm run docgen` from the `wingsdk` folder and check the output, I think that's the easiest way I think to see how docs / bulleted items are rendered.

File: libs/wingsdk/src/http/http.ts [hunk 9]
```diff
@@ -0,0 +1,259 @@
+import { Code, InflightClient } from "../core";
+
+/**
+ * RequestCache-
+ * The cache read-only property that contains the cache mode of the request.
+ * It controls how the request will interact with the browser's HTTP cache.
+
+ */
+export enum RequestCache {
```
Comment (Chriscbr) [inline libs/wingsdk/src/http/http.ts:9]: The docs information here is great. One thing we should update is since we don't know whether the code will be running on a browser or on a cloud serverless function or elsewhere, let's avoid using the term "browser" and instead use "system" or "runtime environment" throughout the descriptions.

File: libs/wingsdk/src/http/http.ts [hunk 10]
```diff
@@ -0,0 +1,259 @@
...
+  /**
+   * Follow all redirects incurred when fetching a resource.
+   */
+  FOLLOW = "follow",
+  /**
+   * Return a network error when a request is met with a redirect.
+   */
+  ERROR = "error",
+}
+/**
```
Comment (Chriscbr) [inline libs/wingsdk/src/http/http.ts:56]: Nit: add a little breathing room between each new class or enum (helps with readability, and in some editors like `vim` they make it easy for users to paginate through definitions if they are separated with newlines)
```suggestion
}

/**
```

File: libs/wingsdk/src/http/http.ts [hunk 11]
```diff
@@ -0,0 +1,259 @@
...
+  /**
+   * The URL of the response.
+   */
+  readonly url: string;
+  /**
+   * A boolean indicating whether the response was successful (status in the range 200 – 299) or not.
+   */
+  readonly ok: boolean;
+  /**
+   * The Headers object associated with the response.
```
Comment (Chriscbr) [inline libs/wingsdk/src/http/http.ts:128]:
```suggestion
   * The map of header information associated with the response.
```

File: libs/wingsdk/src/http/http.ts [hunk 12]
```diff
@@ -0,0 +1,259 @@
...
+ */
+export class Util {
+  /**
+   * @internal
+   */
+  public static _toInflightType(): Code {
+    return InflightClient.forType(__filename, this.name);
+  }
+  /**
+   * Sends an http request of any method.
```
Comment (Chriscbr) [inline libs/wingsdk/src/http/http.ts:158]:
```suggestion
   * Sends an HTTP request to a URL, returning a fulfilled or failed response.
   * @throws Only throws if there is a networking error
```

File: libs/wingsdk/src/http/http.ts [hunk 13]
```diff
@@ -0,0 +1,259 @@
...
+      ...options,
+      method: HttpMethods.DELETE,
+    });
+  }
+
+  private static async _formatResponse(
+    response: globalThis.Response
+  ): Promise<Response> {
+    const headers: Record<string, string> = {};
+    response.headers?.forEach((val: string, key: string) => {
```
[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingsdk/src/http/http.ts:247]: I think this is needed because the Fetch API returns a custom Headers object - looks ok to me.

File: libs/wingsdk/src/http/http.ts [hunk 14]
```diff
@@ -0,0 +1,259 @@
...
+      headers[key] = val;
+    });
```
Comment (Chriscbr) [inline libs/wingsdk/src/http/http.ts:249]: style: Prefer for loop
```suggestion
    // convert Headers object into a plain JS object
    const headers: Record<string, string> = {};
    for (const [key, value] of response.headers.entries()) {
      headers[key] = val;
    }
```

**Comment by tsuf239** [comment]
> > Something to note: we should raise the required node version in our packages/docs from just "18" to "8.13.0", since fetch was experimental until that version. Otherwise users will get a scary warning about it.
> 
> That's a great point. Maybe even raise to the latest LTS 18.16.1?

Cool! Where is it written in the code?

File: examples/tests/valid/api_path_vars.w [hunk 15]
```diff
@@ -1,28 +1,26 @@
...
 
-class Fetch {
-  extern "./api_path_vars.js" inflight get(url: str): Json;
-}
 
 let handler = inflight (req: cloud.ApiRequest): cloud.ApiResponse => {
   return cloud.ApiResponse {
     body: Json.stringify({ user: req.vars.get("name") }),
-    headers: { "content-type": "application/json" },
+    headers: { "content-type": "text/plain" },
```
[the remark this answers is not in the document]
Comment (tsuf239) [reply examples/tests/valid/api_path_vars.w]: restoring now, it was part of the experiments to fix the double stringifying

File: libs/wingsdk/src/http/http.ts (hunk 13 above)
[the remark this answers is not in the document]
Comment (tsuf239) [reply libs/wingsdk/src/http/http.ts]: exactly :)
https://developer.mozilla.org/en-US/docs/Web/API/Headers

File: libs/wingsdk/src/http/http.ts (hunk 6 above)
Comment (tsuf239) [reply libs/wingsdk/src/http/http.ts:108]: it's all taken from here: 
https://developer.mozilla.org/en-US/docs/Web/API/Request/Request

File: libs/wingsdk/src/http/http.ts (hunk 14 above)
Comment (tsuf239) [reply libs/wingsdk/src/http/http.ts]: I wish, but it seems like it doesn't have entries :(
image

File: libs/wingsdk/src/http/http.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (tsuf239) [reply libs/wingsdk/src/http/http.ts:10]: I don't see it generated, I think the folder isn't included, should I do something regarding that? 
(This file is under `src/http`)
https://github.com/winglang/wing/blob/24d91917ea2fd9dd482548dc449936aa81d8854f/libs/wingsdk/.projenrc.ts#L203

File: libs/wingsdk/test/http/http.test.ts (hunk 5 above)
Comment (tsuf239) [reply libs/wingsdk/test/http/http.test.ts:12]: yes, I'll add a wing test :) 
 also, since I'll have to use a real API resource- those can cover also the api's SDK spec tests! #2731

File: libs/wingsdk/src/http/http.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (tsuf239) [reply libs/wingsdk/src/http/http.ts]: it's all taken from there :)
(and from the fetch spec: https://fetch.spec.whatwg.org/#fetch-method)
Is there something specific that needs to be changed?

File: libs/wingsdk/src/http/http.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Chriscbr) [reply libs/wingsdk/src/http/http.ts]: @tsuf239 ah, you'll need to add the module here (sorry its a bit hacky at the moment): https://github.com/winglang/wing/blob/main/apps/jsii-docgen/src/docgen/view/wing-filters.ts

File: examples/tests/sdk_tests/api/delete.w [hunk 16]
```diff
@@ -0,0 +1,38 @@
...
+api.delete("/path", inflight (req: cloud.ApiRequest): cloud.ApiResponse => {
+    assert(req.method == api_DELETE);
+    assert(req.query?.get("all") == "true");
+    assert(req.query?.get("page") == "6");
+    assert(req.path == "/path");
+
+    return cloud.ApiResponse {
+        status: 200,
+        body: req.query?.get("page")
+    };
```
Comment (Chriscbr) [inline examples/tests/sdk_tests/api/delete.w:19]: Should be indented 2 spaces for consistency with our other examples (someday we will have a Wing formatter 🤞)

File: examples/tests/sdk_tests/api/delete.w [hunk 17]
```diff
@@ -0,0 +1,38 @@
+
+bring cloud;
```
Comment (Chriscbr) [inline examples/tests/sdk_tests/api/delete.w:2]: Start examples on first line
```suggestion
bring cloud;
```

File: libs/wingsdk/src/http/http.ts [hunk 18]
```diff
@@ -0,0 +1,267 @@
...
+   */
+  readonly ok: boolean;
+  /**
+   * The map of header information associated with the response.
+   */
+  readonly headers: Record<string, string>;
+  /**
+   * A string represents the body contents.
+   */
+  readonly body: string;
```
Comment (Chriscbr) [inline libs/wingsdk/src/http/http.ts:138]: I wonder if this should be optional, to handle the case where some responses may not have a body (for example, HEAD requests)

**Review by Chriscbr — APPROVED** [review]

File: libs/wingsdk/src/http/http.ts (hunk 18 above)
Comment (tsuf239) [reply libs/wingsdk/src/http/http.ts:138]: image

File: libs/wingsdk/src/http/http.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (tsuf239) [reply libs/wingsdk/src/http/http.ts]: done, thanks @Chriscbr!

File: examples/tests/sdk_tests/api/delete.w [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+bring http;
+
+let http_DELETE = http.HttpMethod.DELETE;
```
Comment (eladb) [inline examples/tests/sdk_tests/api/delete.w:4]: Why is this needed?

**Review by eladb — APPROVED** [review]

File: examples/tests/sdk_tests/api/delete.w (hunk 19 above)
Comment (tsuf239) [reply examples/tests/sdk_tests/api/delete.w:4]: https://winglang.slack.com/archives/C048QCN2XLJ/p1687425563136599
Opened a bug: https://github.com/winglang/wing/issues/3049
I'll add the link to those files
